### PR TITLE
Implement BufRead for BodyReader

### PR DESCRIPTION
**Pull request by bugadani**
This PR allows users to read from a TLS connection without copying. Non-TLS sessions will be able to save some memory by reusing the initial read buffer that was passed in to store the HTTP headers.

### CODE REVIEW

**Comment by bugadani** [comment]
This might be a bit on the complex side as I had to replace the whole concatenating reader concept with a buffering one. But I also couldn't use buffered io because it doesn't support preloaded content that we have after reading headers.

The idea is that, if we have a buffered connection (via embedded-tls), we use its buffer. Otherwise, we read into the rx buffer the user provided while sending the request. We use these buffers as the backing storage of `BufRead`.

I've reimplemented the regular reads in terms of this buffering. This has the downside of one more copy for normal reads. I guess this could be improved by flushing the loaded buffer, and reading directly into the output if we don't have anything currently loaded.

**Comment by bugadani** [comment]
> I've reimplemented the higher-level regular reads in terms of this buffering. This means that the normal read calls may copy twice (once to the internal buffer, once to the output), which isn't ideal.

@lulf @rmja do you have opinions on this? I can resolve this by some not-so-elegant code around the body readers and the new buffered reader here (i.e. by introducing a buffered_len() fn that I could branch on and read into the output directly if empty).

**Review by lulf — APPROVED** [review]
I haven't used these new APIs that much so I'll let @rmja comment as well, but I think it looks good.

**Comment by bugadani** [comment]
> I haven't used these new APIs

The question is more that this PR makes the old API slightly slower :)

**Comment by rmja** [comment]
Is it correctly understood that the underlying non-tls connection now must implement `BufRead`? Could this be more a conditional constraint such that the `BodyReader` implemented `BufRead` if the underlying connection did?

**Comment by bugadani** [comment]
> Is it correctly understood that the underlying non-tls connection now must implement BufRead?

No and if that's currently the case I did something wrong. This PR is supposed to emulate BufRead by reading into the provided rx buffer.

Unfortunately, I don't think it's possible to specialize for a BufRead TcpConnection.

File: src/response.rs [hunk 1]
```diff
@@ -32,13 +31,54 @@ where
...
+            .field("header_len", &self.header_len)
+            .field("raw_body_read", &self.raw_body_read)
+            .finish()
+    }
+}
+
 impl<'buf, 'conn, C> Response<'buf, 'conn, C>
 where
-    C: Read,
+    C: Read + Write,
```
Comment (rmja) [inline src/response.rs:77]: Is there any reason why the connection on the response now must be `Write`?

Comment (bugadani) [reply src/response.rs:77]: Yes, unfortunately the `C` in `HttpConnection<'conn, C>` is `Read + Write` and I think it goes up to `TcpConnect::Connection` being `Read+Write`.

Comment (bugadani) [reply src/response.rs]: Heh, I guess there isn't any reason. Good leading question, it took me way too long to figure out, but whatever. The result is even simpler than before the PR.

**Pull request merged**